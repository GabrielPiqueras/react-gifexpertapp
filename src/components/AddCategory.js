import React, { useState } from 'react';
import PropTypes from 'prop-types';

/**
 * COMPONENTE ADDCATEGORY
 * 
 * 1. Muestra el estado del componente (inputValue) en la caja de texto.
 * 2. Si cambia el valor de la caja de texto, dicho valor será el nuevo estado.
 * 3. Si se envía el formulario y el texto tiene al menos 2 letras, se añade la categoría al array.
 * 4. Al recibir la función como propiedad que cambia el valor del estado del componente GifExpertApp (setCategories), 
 *    podemos añadir el nuevo valor al array.
 * 5. Se reinicia el input para dejarlo vacío.
 * 
 */
const AddCategory = ({ setCategories }) => {
        
    const [ inputValue, setInputValue ] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault();

        if (inputValue.trim().length > 2) {
            setCategories( cats => [inputValue, ...cats ]);
            setInputValue('');
        }
    }

    return(
        <form onSubmit={ handleSubmit }>
            <input
                onChange={ handleInputChange }
                id='category'
                type='text'
                value={ inputValue }
                placeholder='Search'
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}

export default AddCategory;