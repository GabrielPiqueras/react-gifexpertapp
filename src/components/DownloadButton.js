import React from 'react';
import { ArrowDownTrayIcon } from '@heroicons/react/20/solid'

const DownloadButton = ({ gifUrl, fileName }) => {
    
  const handleDownload = async () => {
    const response = await fetch(gifUrl);
    const blob = await response.blob();
    const url = window.URL.createObjectURL(blob);

    const link = document.createElement('a');
    link.href = url;
    link.download = fileName;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    window.URL.revokeObjectURL(url);
  };

  return (
    <button
      className="flex justify-end items-start ml-auto bg-gray-500 hover:bg-gray-400 text-white font-bold py-2 px-2 border-b-4 border-gray-700 hover:border-gray-500 rounded"
      onClick={handleDownload}
      title='Descargar Gif'
    >
      <ArrowDownTrayIcon className="w-4 h-4"/>
    </button>
  );
}

export default DownloadButton;