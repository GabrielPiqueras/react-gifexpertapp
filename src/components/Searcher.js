import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { MagnifyingGlassIcon } from '@heroicons/react/20/solid'

/**
 * COMPONENTE ADDCATEGORY
 * 
 * 1. Muestra el estado del componente (inputValue) en la caja de texto.
 * 2. Si cambia el valor de la caja de texto, dicho valor será el nuevo estado.
 * 3. Si se envía el formulario y el texto tiene al menos 2 letras, se añade la categoría al array.
 * 4. Al recibir la función como propiedad que cambia el valor del estado del componente GifSearcher (setPrompt), 
 *    podemos añadir el nuevo valor al array.
 * 5. Se reinicia el input para dejarlo vacío.
 * 
 */
export const Searcher = ({ setPrompt }) => {
        
    const [ inputValue, setInputValue ] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault();

        if (inputValue.trim().length > 2) {
            setPrompt(inputValue);
            setInputValue('');
        }
    }

    return(
        <form onSubmit={ handleSubmit }>

            <div className="mt-12 mb-4 flex justify-center">
                <input
                    autoFocus 
                    id='category'
                    type='text'
                    value={ inputValue }
                    onChange={ handleInputChange }
                    style={{width: '35%'}}
                    className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline w-64 h-12 mr-4"
                    placeholder="Gifs de..."
                />
                <button
                    className="bg-blue-500 hover:bg-blue-400 text-white font-bold py-2 px-4 border-b-4 border-blue-700 hover:border-blue-500 rounded flex items-center"
                    onClick={ handleSubmit }
                >
                    <MagnifyingGlassIcon className="w-5 h-5 mr-1"/> Buscar
                </button>
            </div>
        </form>
    )
}

Searcher.propTypes = {
    setPrompt: PropTypes.func.isRequired,
}