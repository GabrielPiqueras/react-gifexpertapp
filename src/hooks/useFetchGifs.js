
/* Esto es un custom hook, si no vamos a devolver JSX no es necesario que importemos React (el general) */
/* Otra cosa es que necesitemos usar useState u otros */

/* Estos hooks también tienen un estado y podría indicarle a otros componentes que lo utilicen cuando deben renderizarse
ya que algo cambió */

import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (prompt) => {
    
    /* El estado será un objeto, con un array para la data y un loading a 'true',
    que indicará que cuando se utilice por primera vez este hook "si que cargue", por defecto */
    
    const [ state, setState ] = useState({
        data: [],
        loading: true
    });

    // Uso useEffect para controlar que solo cargue una vez
    useEffect(() => {
        getGifs(prompt).then(gifs => {

            setState({
                data: gifs,
                loading: false
            });
        });
    }, []);

    return state; // { data: [], loading: true }
}
