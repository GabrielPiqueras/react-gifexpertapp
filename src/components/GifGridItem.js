import React from 'react'
import PropTypes from 'prop-types';
import DownloadButton from './DownloadButton';

export const GifGridItem = ({ id, title, url }) => {

  const convertToSlug =(title)  => {
    const slug = title
      .toLowerCase()
      .replace(/[^\w\s-]/g, '') // Elimina caracteres especiales
      .replace(/\s+/g, '-') // Reemplaza espacios en blanco con guiones
      .replace(/--+/g, '-'); // Elimina guiones consecutivos
    
    return slug;
  }

  return (
    <>
      {/* Nueva */}
      <div className="bg-white max-w-sm rounded overflow-hidden shadow-lg card animate__animated animate__fadeIn m-4 w-1/1 sm:w-1/2 md:w-1/3 lg:w-1/4 xl:w-1/3">
        <img className="w-full" src={ url } alt={ title} />
        <div className="py-4">
          <div className="pl-6 pr-4 font-bold text-xl mb-2 flex justify-end items-start">
            { title }
            <DownloadButton gifUrl={ url } fileName={ convertToSlug(title) } />
            
            </div>
          
        </div>
        <div>
          <p className="px-6 text-gray-700 text-base">
            Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus
            quia, nulla! Maiores et perferendis eaque, exercitationem praesentium
            nihil.
          </p>
        </div>
        <div className="px-6 pt-4 pb-2">
          <span className="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2">
            #gif
          </span>
          <span className="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2">
            #hastag
          </span>
          <span className="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2">
            #other
          </span>

            

          {/* <button class="bg-blue-500 hover:bg-blue-400 text-white font-bold py-2 px-4 border-b-4 border-blue-700 hover:border-blue-500 rounded">
            Descargar GIF
          </button> */}
         
        </div>
      </div>
    </>
  )
}

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
}