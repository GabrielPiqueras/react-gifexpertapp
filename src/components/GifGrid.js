import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

import PropTypes from 'prop-types';

// import { getGifs } from '../helpers/getGifs';

/**
 * COMPONENTE GIFGRID
 * 
 * - Muestra los gifs según el término buscado en giphy, son mostrados en un div con sus títulos.
 * 
 * 1. El estado del componente será un array (vacío por defecto) donde se guardarán los gifs
 * 2. Los gifs se obtienen usando la Api de Giphy
 * 3. Se buscan los gifs poniendo mi ApiKey de la plataforma, el límite de gifs y el término a buscar, que será la categoría
 *    recibida por parámetro. Dichas categorías se guardan en un array del componente principal.
 * 4. Al recibir los gifs en la variable data, la mapeo para extraer el id, titulo e imagen de cada gif.
 * 5. Una vez hecho el mapeo, se pone dicho array como valor del estado.
 * 6. Usamos el Hook useEffect() para que solo haga la petición una vez al principio, y no si se recarga el componente.
 * 7. Pinto los gifs del array en html, con sus titulos e imagenes.
 */

export const GifGrid = ({ category }) => {

  // Uso mi hook personalizado que trae los gifs
  const { data: gifs, loading } = useFetchGifs(category);

  return (
    <>
      <h3>{ category }</h3>

      {/* Más práctico de mostrar */}
      { loading && <p>Cargando...</p> } 
      {/* { loading ? <p>Cargando...</p>: <p>Otro mensaje.</p> } */}

      <div className='card-grid'>
          {
            // Usando el {...} sería como enviar el id, title y url como propiedades por separado (otra forma de hacerlo) en vez de pasarlo dentro de img
            gifs.map(gif =>
              <GifGridItem key={ gif.id } {...gif} />
              // <GifGridItem key={ gif.id } img={ gif } />
            )
          }
      </div>
    </>
  )
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired, 
} 
