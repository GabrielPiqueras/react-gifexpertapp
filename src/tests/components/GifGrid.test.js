import React from 'react';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import '@testing-library/jest-dom';
import { useFetchGifs } from '../../hooks/useFetchGifs';

// jest.mock() finge la llamada al archivo para suponer o controlar la información que va a devolver */
jest.mock('../../hooks/useFetchGifs');

// Nos dará un error en consola, diciendo que no podemos 
// hacer la desestructuración de data 
// El fallo dará al renderizar mi componente <GifGrid />, es buena señal,
// porque quiere decir que useFetchGifs() ahora se está utilizando 



describe('Pruebas en <GifGrid />', () => {

    const category = 'One Piece';

    test('El componente <GifGrid /> debe renderizarse correctamente', () => {

        // Para poder "falsear" los datos recibidos del useFetchGifs
        // usaremos mockReturnValue(), hace que cuando se llame el useFetchGifs dentro
        // de <GifGrid /> lo que va a hacer es regresar este valor.
        
        // Como devolverá un objeto en este caso, ponemos las {}
        // Como el estado inicial es que devuelva la data vacía y el loading true,
        // ponemos lo que esperamos que devuelva:
            
       useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow( <GifGrid category={ category } /> );
        expect( wrapper ).toMatchSnapshot();
    });

    test('El componente <GifGrid /> debe mostrar items cuando se cargan las imágenes', () => {
        
        // En este caso espero recibir los datos con esta estructura
        // Nos inventamos unos datos falsos para comprobar que la estructura coincide
        const gifs = [
            {
                id: 'ABC',
                url: 'https://localhost/cualquier/cosa/.jpg',
                title: 'Cualquier cosa'
            }, {
                id: 'XYZ',
                url: 'https://localhost/cualquier/cosa2/.jpg',
                title: 'Cualquier cosa 2'
            }
        ];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        // Renderizamos 
        const wrapper = shallow( <GifGrid category={ category } /> );
        
        expect(wrapper).toMatchSnapshot(); // El componente se renderiza bien
        expect(wrapper.find('p').exists()).toBe(false); // No existe el párrafo de "cargando" porque ya cargaron los gifs
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length); // Habrá tantos GifGridItem como objetos tenga
    });

});