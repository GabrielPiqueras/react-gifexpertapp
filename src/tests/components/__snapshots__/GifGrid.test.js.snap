// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> El componente <GifGrid /> debe mostrar items cuando se cargan las im√°genes 1`] = `
<Fragment>
  <h3>
    One Piece
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="https://localhost/cualquier/cosa/.jpg"
    />
    <GifGridItem
      id="XYZ"
      key="XYZ"
      title="Cualquier cosa 2"
      url="https://localhost/cualquier/cosa2/.jpg"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en <GifGrid /> El componente <GifGrid /> debe renderizarse correctamente 1`] = `
<Fragment>
  <h3>
    One Piece
  </h3>
  <p>
    Cargando...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
